[Sema] Add specialization constraints for func and variable types, then diagnose w/fixes.

gregomni opened the pull request:
Similar logic to the diagnoses already happening in visitUnresolvedSpecializeExpr in CSGen, but checking again after there are no more type variables resolves #74857.

gregomni commented:
@swift-ci Please smoke test.

AnthonyLatsis reviewed (commented):
Looks good, but I think we should be diagnosing more cases here, perhaps anything expect macro specializations and a `DotSyntaxBaseIgnoredExpr` subsexpression, which is this sort of case:
```swift
extension Bool {
  typealias A<T> = Bool
}

func foo(_ meta: Bool.Type) {
  let _ = meta.A<Bool>.self
}
```

gregomni commented on lib/Sema/CSApply.cpp:
```diff
@@ -3225,8 +3225,18 @@ namespace {
+      // recheck for validity.
+      auto subExpr = expr->getSubExpr();
+      auto ty = subExpr->getType();
+      if (ty && !ty->is<AnyMetatypeType>()) {
```
Fixed!

gregomni commented on test/Sema/generic-arg-list.swift:
```diff
@@ -0,0 +1,26 @@
+
+  // Now also Error
+  let _ = 0.foo<Int>() // expected-error {{cannot specialize a non-generic definition}}
+    // expected-note@-1 {{while parsing this '<' as a type parameter bracket}}
```
Fixed!

gregomni commented on test/Sema/generic-arg-list.swift:
```diff
@@ -0,0 +1,26 @@
+func test(i: Int) {
+  // Error
+  let _ = i.foo<Int>() // expected-error {{cannot explicitly specialize a generic function}}
+    // expected-note@-1 {{while parsing this '<' as a type parameter bracket}}
```
Fixed!

gregomni commented:
@swift-ci Please smoke test.

gregomni commented:
I think I've resolved the issues and solidified this. Now delays checking of all apparent function types and diagnoses them all afterwards in CSApply (or if we would otherwise get an unbound generic parameter error -- there).

Also, could you give me the gist of the request for this radar in `test/decl/ext/specialize.swift`: rdar://111059036 
Is there a diagnostic wanted here specifically for trying to specialize `Self`? (Now easy to add, if so.)

xedin reviewed (changes requested):
A more reliable way to implement this would be to introduce the specialization constraint during CSGen regardless of what the base type is and use it to introduce fixes when types are sufficiently resolved.

Diagnosing something in CSApply is in general too late because it means that the solver produced an invalid solution which shouldn't happen for things that could be checked during solving.

gregomni commented:
> A more reliable way to implement this would be to introduce the specialization constraint during CSGen regardless of what the base type is and use it to introduce fixes when types are sufficiently resolved.

Done! I left the check in CSGen for known non-function types because diagnoses pointing at the type-parameter bracket is better for confusing situations like `Parse/generic_disambiguation.swift:29`: 
```
var a, b, c, d : Int
(a < b, c > (d))
```
Where if you create the constraint you also get extraneous errors about `b` and `c` not being known types. But for everything else it now just creates the specialization constraint and diagnoses via the fix for it.

gregomni commented:
@swift-ci Please smoke test.

xedin commented:
Thanks! I will take a look on monday!

gregomni commented:
Smoke test failed on Linux because a note was slightly different (different matching types on operator '<'). Replaced with  substring of emitted note.

gregomni commented:
@swift-ci Please smoke test.

Commit:
Improve diagnoses of generic specializations

Always add constraints, find fixes during simplify.
New separate fix for allow generic function specialization.
Improve parse heuristic for isGenericTypeDisambiguatingToken.

gregomni commented:
@swift-ci Please smoke test.

hamishknight commented on lib/Sema/CSGen.cpp:
```diff
@@ -1937,40 +1937,30 @@ namespace {
+        // FIXME: If the base type is a type variable, constrain it to a
+        // metatype of a bound generic type.
+        auto &de = CS.getASTContext().Diags;
         de.diagnose(expr->getSubExpr()->getLoc(),
-                    diag::cannot_explicitly_specialize_generic_function);
+                    diag::not_a_generic_definition);
         de.diagnose(expr->getLAngleLoc(),
                     diag::while_parsing_as_left_angle_bracket);
         return Type();
```
If you want, while here it would be nice to get rid of this diagnostic logic; CSGen should never emit diagnostics since it can be run multiple times for conjunction elements. Moreover, CSGen should never fail since it prevents a valid solution from being formed, meaning we emit an unhelpful "unable to infer" diagnostic for e.g:

```swift
func foo(_ x: Int) {
  _ = { // error: unable to infer closure type without a type annotation
    _ = x<String> // error: cannot specialize a non-generic definition
  }
}
```

Instead we should record a fix, maybe we can rely on the diagnostic logic in the specialization constraint for that.

gregomni replied:
I can look at that again -- I tried removing this logic and it leads to some regressions (quoted from above: "I left the check in CSGen for known non-function types because diagnoses pointing at the type-parameter bracket is better for confusing situations like `Parse/generic_disambiguation.swift:29`").  But maybe work-around-able.

gregomni replied:
Yeah, `addSpecializationConstraint` immediately tries to resolve the typereprs in the list, so in the sort of case we are talking about, if you tried to add the constraint you'd get a `cannot find type 'b' in scope` diagnosis immediately from CSGen.

hamishknight replied:
Hmm I see, you could always record the fix directly in CSGen to bypass that

gregomni replied:
Actually, the fallout is not bad from that. So long as you get the note telling you the compiler parsed this as a specialization, the can't-find-type error is fine, and this does improve your example.

Added this as a second commit so you can see the differences, will squash if/when approved.

gregomni replied:
(This fixes the cases where there are conjunctions involved in trying to specialize a non-generic type. It does NOT fix when there are conjunctions involved in trying to specialize with a type in the list that doesn't exist).

gregomni commented:
Also added a commit that improves parsing to avoid many cases of unary postfix '>'.

@swift-ci Please smoke test.

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -1937,40 +1937,23 @@ namespace {
-      if (baseTy->isTypeVariableOrMember()) {
-        return baseTy;
+      if (AnyMetatypeType *meta = baseTy->getAs<AnyMetatypeType>()) {
+        specializeTy = meta->getInstanceType();
```
We have a way to look through metatypes on `TypeBase` - `getMetatypeInstanceType()`.

gregomni replied:
Fixed!

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -1937,40 +1937,23 @@ namespace {
+        auto &de = CS.getASTContext().Diags;
         de.diagnose(expr->getLAngleLoc(),
                     diag::while_parsing_as_left_angle_bracket);
         return Type();
```
I'd prefer if we didn't diagnose and return `Type()` and subsequently fail the constraint generation because that means that the expression that contains this specialization won't get any code completion...

xedin replied:
In fact we should probably use `addUnsolvedConstraint` here to delay its simplification until solving begins.

gregomni replied:
Now doesn't fail constraint generation, replaces missing types with `PlaceholderType`. I don't think there's any need to delay with `addUnsolvedConstraint`?

xedin commented on test/Parse/enum_element_pattern_swift4.swift:
```diff
@@ -9,10 +9,11 @@ enum E {
   static func testE(e: E) {
     switch e {
-    case A<UndefinedTy>(): // expected-error {{cannot specialize a non-generic definition}}
+    case A<UndefinedTy>(): // expected-error {{cannot find type 'UndefinedTy' in scope}}
```
Maybe this is something that could be addressed during pre-check which would set a placeholder type for `UndefinedTy` declaration in this context?

gregomni replied:
Yep, fixed!

xedin commented on lib/Sema/CSDiagnostics.h:
```diff
@@ -3111,12 +3111,13 @@ class InvalidMemberReferenceWithinInitAccessor final
 /// \endcode
 class ConcreteTypeSpecialization final : public FailureDiagnostic {
   Type ConcreteType;
+  ValueDecl *Decl;
```
I think actually a `PointerUnion<TypeBase *, ValueDecl *>` since like either or based on the `::diagnoseAsError()` implementation.

gregomni replied:
True. There isn't actually a need for the `ValueDecl` in generating the diagnosis message either, so changed this to check whether the decl is a generic function up front, and then only store the ConcreteType again, and if it is nullptr, that means the other diagnosis.

xedin commented on lib/Parse/ParseType.cpp:
```diff
@@ -1474,7 +1474,9 @@ static bool isGenericTypeDisambiguatingToken(Parser &P) {
-    return false;
+    // If this is the end of the expr (wouldn't match parseExprSequenceElement),
+    // prefer generic type list over an illegal unary postfix '>' operator.
+    return tok.isKeyword() && tok.getKind() != tok::kw_try;
```
@rintaro could you take a look at this please?

rintaro replied:
`tok.isKeyword()` is usually not enough for detecting "the token is not related to the current production" because many decls starts with a non-keyword.  e.g.
```swift
let a = Foo<Bar>
open class Cls { ... }
```

`P.isStartOfSwiftDecl() || P.isStartOfStmt()` might be better.

gregomni commented:
Did all the Sema review feedback. Will look at Parse when I get a chance -- or would it be easier if I did a separate PR for that part with a different reviewer?

@swift-ci Please smoke test.

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -1916,61 +1917,39 @@ namespace {
+        if (result->hasError()) {
+          auto &ctxt = CS.getASTContext();
+          auto *repr = new (ctxt) PlaceholderTypeRepr(specializationArg->getLoc());
+          result = PlaceholderType::get(ctxt, repr);
```
Instead of create a placeholder repr placeholder type should use original `specializationArg`

gregomni replied:
Fixed

gregomni replied:
It looks like `replaceInferableTypesWithTypeVars` does this same pattern of creating a `PlaceholderTypeRepr`, so I think this should be left as is.

xedin replied:
That method is actually going the other way - it detects if there are placeholders in a `Type` and "opens" them. I think we should expand from PlaceholderTypeRepr to `TypeRepr` in `PlaceholderType` and locator element. But we can do that in a separate PR.

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -1916,61 +1917,39 @@ namespace {
+      if (addSpecializationConstraint(overloadLocator, baseTy->getMetatypeInstanceType(),
+                                      expr->getUnresolvedParams())) {
+        // Diagnosis of failed type lookup - add a bit of context
+        auto &de = CS.getASTContext().Diags;
```
I still think we should delay simplification here to let "normal" solving fail and re-attempt during diagnostic mode.

gregomni replied:
Sorry, I think you'll have to be more explicit, I'm not sure I understand.

`CS.addConstraint()` happens inside of `addSpecializationConstraint()`. If this is changed to be `CS.addUnsolvedConstraint()`, then the solve for many of these cases will type check while leaving the specialization constraint in the Inactive list the whole time (because they really are extraneous), and then fail due to remaining inactive constraints (CSStep.cpp:482).

The if statement here was just for whether or not to add a note, the constraint is added no matter what. But I'll clean that part up.

xedin replied:
To be precise my suggestion was to remove the result type for `addSpecializationConstraint`, inside of that method: create a new `ExplicitGenericArguments` constraint, use `addUnsolvedConstraint` + `activateConstraint` (usually `activateConstraint` is not necessary but as far as I understand the main problem here that sometimes there are no type variables here for solver to bind and subsequently re-evaluate the constraint).

One trick to avoid explicit activation would be to allocate a new type variable and use an unsolved `Equal` constraint to connect it to the `boundType`, type variable itself should be used in place of `boundType` in an unsolved `ExplicitGenericArguments` constraint. This way the solver would pick the type variable and bind it during solving which activates all of the constraints that reference it.

gregomni replied:
Gotcha. Thanks for expanding.

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -4082,10 +4061,8 @@ namespace {
-              expr->getGenericArgs()))
-          return Type();
+        addSpecializationConstraint(CS.getConstraintLocator(expr), macroRefType,
+                                    expr->getGenericArgs());
```
Should use `addUnsolvedConstraint` here as well instead of completely ignoring the result.

gregomni replied:
See above, the result was just for adding a note, but I've change this now.

xedin commented on lib/Sema/CSFix.cpp:
```diff
@@ -2599,7 +2599,15 @@ bool AllowConcreteTypeSpecialization::diagnose(const Solution &solution,
+  if (isa<AbstractFunctionDecl>(decl)) {
+    auto genericContext = decl->getAsGenericContext();
+    if (genericContext && genericContext->getGenericParams()) {
+      concreteTy = nullptr;
```
Hm, I think this is actually a regression over what you had before but I think we need here is two fixes and two district diagnostics: one for a concrete type and one for a generic function. The latter doesn't have to carry the type but could benefit from having a declaration since we could add a note that points out where the function/method is declared.

gregomni replied:
Yeah, I should've made it a separate fix type the whole time, and these were bad shortcuts.

xedin reviewed (commented):
I think we are super close here, left a few more comments inline, also we are focusing on methods/functions but what about subscripts? It would be good to add a few test-cases for them as well.

gregomni commented:
Okay! Feedback commit, round 2. Unfortunately, I have some questions about what you're asking for, so this isn't likely to be final.

gregomni commented:
Ok, I think maybe done. Squashed changes so far.

@rintaro's suggestion of how to do the Parse change works perfectly.
The constraints are now added unsolved and then activated.

@swift-ci Please smoke test.

rintaro commented:
@swift-ci Please smoke test

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -9338,7 +9338,12 @@ bool InvalidMemberReferenceWithinInitAccessor::diagnoseAsError() {
 
 bool ConcreteTypeSpecialization::diagnoseAsError() {
-  emitDiagnostic(diag::not_a_generic_type, ConcreteType);
+  emitDiagnostic(diag::not_a_generic_type, resolveType(ConcreteType));
```
Nit: I'd prefer the resolveType happened during construction of the diagnostic, it's usually safer that way.

gregomni replied:
Done. That was an accidental leftover from the changes along the way.

xedin reviewed (approved):
Thank you very much for all the work here!

gregomni commented:
@swift-ci Please smoke test.
